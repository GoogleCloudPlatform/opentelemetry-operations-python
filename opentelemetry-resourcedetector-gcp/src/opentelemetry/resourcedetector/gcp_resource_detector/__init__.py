# Copyright 2021 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
import os

import requests
from opentelemetry.context import attach, detach, set_value
from opentelemetry.sdk.resources import Resource, ResourceDetector

_GCP_METADATA_URL = (
    "http://metadata.google.internal/computeMetadata/v1/?recursive=true"
)
_GCP_METADATA_URL_HEADER = {"Metadata-Flavor": "Google"}

logger = logging.getLogger(__name__)


def _get_google_metadata_and_common_attributes():
    token = attach(set_value("suppress_instrumentation", True))
    all_metadata = requests.get(
        _GCP_METADATA_URL, headers=_GCP_METADATA_URL_HEADER
    ).json()
    detach(token)
    common_attributes = {
        "cloud.account.id": all_metadata["project"]["projectId"],
        "cloud.provider": "gcp",
        "cloud.zone": all_metadata["instance"]["zone"].split("/")[-1],
    }
    return common_attributes, all_metadata


def get_gce_resources():
    """Resource finder for common GCE attributes

    See: https://cloud.google.com/compute/docs/storing-retrieving-metadata
    """
    (
        common_attributes,
        all_metadata,
    ) = _get_google_metadata_and_common_attributes()
    common_attributes.update(
        {
            "host.id": all_metadata["instance"]["id"],
            "gcp.resource_type": "gce_instance",
        }
    )
    return common_attributes


def get_gke_resources():
    """Resource finder for GKE attributes"""

    if os.getenv("KUBERNETES_SERVICE_HOST") is None:
        return {}

    (
        common_attributes,
        all_metadata,
    ) = _get_google_metadata_and_common_attributes()

    container_name = os.getenv("CONTAINER_NAME")
    if container_name is not None:
        common_attributes["container.name"] = container_name

    # Fallback to reading namespace from a file is the env var is not set
    pod_namespace = os.getenv("NAMESPACE")
    if pod_namespace is None:
        try:
            with open(
                "/var/run/secrets/kubernetes.io/serviceaccount/namespace", "r"
            ) as namespace_file:
                pod_namespace = namespace_file.read().strip()
        except FileNotFoundError:
            pod_namespace = ""

    common_attributes.update(
        {
            "k8s.cluster.name": all_metadata["instance"]["attributes"][
                "cluster-name"
            ],
            "k8s.namespace.name": pod_namespace,
            "k8s.pod.name": os.getenv("POD_NAME", os.getenv("HOSTNAME", "")),
            "host.id": all_metadata["instance"]["id"],
            "gcp.resource_type": "gke_container",
        }
    )
    return common_attributes


def get_cloudrun_resources():
    """Resource finder for Cloud Run attributes"""

    if os.getenv("K_CONFIGURATION") is None:
        return {}

    (
        common_attributes,
        all_metadata,
    ) = _get_google_metadata_and_common_attributes()

    faas_name = os.getenv("K_SERVICE")
    if faas_name is not None:
        common_attributes["faas.name"] = str(faas_name)

    faas_version = os.getenv("K_REVISION")
    if faas_version is not None:
        common_attributes["faas.version"] = str(faas_version)

    common_attributes.update(
        {
            "cloud.platform": "gcp_cloud_run",
            "cloud.region": all_metadata["instance"]["region"].split("/")[-1],
            "faas.instance": all_metadata["instance"]["id"],
            "gcp.resource_type": "cloud_run",
        }
    )
    return common_attributes


def get_cloudfunctions_resources():
    """Resource finder for Cloud Functions attributes"""

    if os.getenv("FUNCTION_TARGET") is None:
        return {}

    (
        common_attributes,
        all_metadata,
    ) = _get_google_metadata_and_common_attributes()

    faas_name = os.getenv("K_SERVICE")
    if faas_name is not None:
        common_attributes["faas.name"] = str(faas_name)

    faas_version = os.getenv("K_REVISION")
    if faas_version is not None:
        common_attributes["faas.version"] = str(faas_version)

    common_attributes.update(
        {
            "cloud.platform": "gcp_cloud_functions",
            "cloud.region": all_metadata["instance"]["region"].split("/")[-1],
            "faas.instance": all_metadata["instance"]["id"],
            "gcp.resource_type": "cloud_functions",
        }
    )
    return common_attributes


# Order here matters. Since a GKE_CONTAINER is a specialized type of GCE_INSTANCE
# We need to first check if it matches the criteria for being a GKE_CONTAINER
# before falling back and checking if its a GCE_INSTANCE.
# This list should be sorted from most specialized to least specialized.
_RESOURCE_FINDERS = [
    ("gke_container", get_gke_resources),
    ("cloud_run", get_cloudrun_resources),
    ("cloud_functions", get_cloudfunctions_resources),
    ("gce_instance", get_gce_resources),
]


class NoGoogleResourcesFound(Exception):
    pass


class GoogleCloudResourceDetector(ResourceDetector):
    def __init__(self, raise_on_error=False):
        super().__init__(raise_on_error)
        self.cached = False
        self.gcp_resources = {}

    def detect(self) -> "Resource":
        if not self.cached:
            self.cached = True
            for resource_type, resource_finder in _RESOURCE_FINDERS:
                try:
                    found_resources = resource_finder()
                # pylint: disable=broad-except
                except Exception as ex:
                    logger.warning(
                        "Exception %s occured attempting %s resource detection",
                        ex,
                        resource_type,
                    )
                    found_resources = None
                if found_resources:
                    self.gcp_resources = found_resources
                    break
        if not self.gcp_resources:
            raise NoGoogleResourcesFound()
        return Resource(self.gcp_resources)
