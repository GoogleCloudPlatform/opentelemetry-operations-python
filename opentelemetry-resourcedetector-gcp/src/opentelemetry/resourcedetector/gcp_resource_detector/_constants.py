# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# TODO: use opentelemetry-semantic-conventions package for these constants once it has
# stabilized. Right now, pinning an unstable version would cause dependency conflicts for
# users so these are copied in.
class ResourceAttributes:
    AWS_EC2 = "aws_ec2"
    CLOUD_ACCOUNT_ID = "cloud.account.id"
    CLOUD_AVAILABILITY_ZONE = "cloud.availability_zone"
    CLOUD_PLATFORM_KEY = "cloud.platform"
    CLOUD_PROVIDER = "cloud.provider"
    CLOUD_REGION = "cloud.region"
    GCP_COMPUTE_ENGINE = "gcp_compute_engine"
    GCP_KUBERNETES_ENGINE = "gcp_kubernetes_engine"
    HOST_ID = "host.id"
    HOST_NAME = "host.name"
    HOST_TYPE = "host.type"
    K8S_CLUSTER_NAME = "k8s.cluster.name"
    K8S_CONTAINER_NAME = "k8s.container.name"
    K8S_NAMESPACE_NAME = "k8s.namespace.name"
    K8S_NODE_NAME = "k8s.node.name"
    K8S_POD_NAME = "k8s.pod.name"
    SERVICE_INSTANCE_ID = "service.instance.id"
    SERVICE_NAME = "service.name"
    SERVICE_NAMESPACE = "service.namespace"
    FAAS_INSTANCE = "faas.instance"
    FAAS_NAME = "faas.name"


AWS_ACCOUNT = "aws_account"
AWS_EC2_INSTANCE = "aws_ec2_instance"
CLUSTER_NAME = "cluster_name"
CONTAINER_NAME = "container_name"
GCE_INSTANCE = "gce_instance"
GENERIC_NODE = "generic_node"
GENERIC_TASK = "generic_task"
INSTANCE_ID = "instance_id"
JOB = "job"
K8S_CLUSTER = "k8s_cluster"
K8S_CONTAINER = "k8s_container"
K8S_NODE = "k8s_node"
K8S_POD = "k8s_pod"
LOCATION = "location"
NAMESPACE = "namespace"
NAMESPACE_NAME = "namespace_name"
NODE_ID = "node_id"
NODE_NAME = "node_name"
POD_NAME = "pod_name"
REGION = "region"
TASK_ID = "task_id"
ZONE = "zone"
UNKNOWN_SERVICE_PREFIX = "unknown_service"
