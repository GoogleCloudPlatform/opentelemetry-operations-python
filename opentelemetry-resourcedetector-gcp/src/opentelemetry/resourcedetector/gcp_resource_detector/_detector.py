# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import Mapping

from opentelemetry.resourcedetector.gcp_resource_detector import (
    _gce,
    _gke,
    _metadata,
)
from opentelemetry.resourcedetector.gcp_resource_detector._constants import (
    ResourceAttributes,
)
from opentelemetry.sdk.resources import Resource, ResourceDetector
from opentelemetry.util.types import AttributeValue


class GoogleCloudResourceDetector(ResourceDetector):
    def detect(self) -> Resource:
        if not _metadata.is_available():
            return Resource.get_empty()

        if _gke.on_gke():
            return _gke_resource()
        if _gce.on_gce():
            return _gce_resource()

        return Resource.get_empty()


def _gke_resource() -> Resource:
    zone_or_region = _gke.availability_zone_or_region()
    zone_or_region_key = (
        ResourceAttributes.CLOUD_AVAILABILITY_ZONE
        if zone_or_region.type == "zone"
        else ResourceAttributes.CLOUD_REGION
    )
    return _make_resource(
        {
            ResourceAttributes.CLOUD_PLATFORM_KEY: ResourceAttributes.GCP_KUBERNETES_ENGINE,
            zone_or_region_key: zone_or_region.value,
            ResourceAttributes.K8S_CLUSTER_NAME: _gke.cluster_name(),
            ResourceAttributes.HOST_ID: _gke.host_id(),
        }
    )


def _gce_resource() -> Resource:
    zone_and_region = _gce.availability_zone_and_region()
    return _make_resource(
        {
            ResourceAttributes.CLOUD_PLATFORM_KEY: ResourceAttributes.GCP_COMPUTE_ENGINE,
            ResourceAttributes.CLOUD_AVAILABILITY_ZONE: zone_and_region.zone,
            ResourceAttributes.CLOUD_REGION: zone_and_region.region,
            ResourceAttributes.HOST_TYPE: _gce.host_type(),
            ResourceAttributes.HOST_ID: _gce.host_id(),
            ResourceAttributes.HOST_NAME: _gce.host_name(),
        }
    )


def _make_resource(attrs: Mapping[str, AttributeValue]) -> Resource:
    return Resource.create(
        {
            ResourceAttributes.CLOUD_PROVIDER: "gcp",
            ResourceAttributes.CLOUD_ACCOUNT_ID: _metadata.get_metadata()[
                "project"
            ]["projectId"],
            **attrs,
        }
    )
