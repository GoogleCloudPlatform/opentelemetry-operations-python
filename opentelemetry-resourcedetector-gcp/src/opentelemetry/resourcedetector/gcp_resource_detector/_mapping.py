# Copyright 2023 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
from dataclasses import dataclass
from typing import Dict, Mapping, Optional, Tuple

from opentelemetry.resourcedetector.gcp_resource_detector import _constants
from opentelemetry.resourcedetector.gcp_resource_detector._constants import (
    ResourceAttributes,
)
from opentelemetry.sdk.resources import Attributes, Resource


class MapConfig:
    otel_keys: Tuple[str, ...]
    """
    OTel resource keys to try and populate the resource label from. For entries with multiple
    OTel resource keys, the keys' values will be coalesced in order until there is a non-empty
    value.
    """

    fallback: str
    """If none of the otelKeys are present in the Resource, fallback to this literal value"""

    def __init__(self, *otel_keys: str, fallback: str = ""):
        self.otel_keys = otel_keys
        self.fallback = fallback


# Mappings of GCM resource label keys onto mapping config from OTel resource for a given
# monitored resource type. Copied from Go impl:
# https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/v1.8.0/internal/resourcemapping/resourcemapping.go#L51
MAPPINGS = {
    _constants.GCE_INSTANCE: {
        _constants.ZONE: MapConfig(ResourceAttributes.CLOUD_AVAILABILITY_ZONE),
        _constants.INSTANCE_ID: MapConfig(ResourceAttributes.HOST_ID),
    },
    _constants.K8S_CONTAINER: {
        _constants.LOCATION: MapConfig(
            ResourceAttributes.CLOUD_AVAILABILITY_ZONE,
            ResourceAttributes.CLOUD_REGION,
        ),
        _constants.CLUSTER_NAME: MapConfig(
            ResourceAttributes.K8S_CLUSTER_NAME
        ),
        _constants.NAMESPACE_NAME: MapConfig(
            ResourceAttributes.K8S_NAMESPACE_NAME
        ),
        _constants.POD_NAME: MapConfig(ResourceAttributes.K8S_POD_NAME),
        _constants.CONTAINER_NAME: MapConfig(
            ResourceAttributes.K8S_CONTAINER_NAME
        ),
    },
    _constants.K8S_POD: {
        _constants.LOCATION: MapConfig(
            ResourceAttributes.CLOUD_AVAILABILITY_ZONE,
            ResourceAttributes.CLOUD_REGION,
        ),
        _constants.CLUSTER_NAME: MapConfig(
            ResourceAttributes.K8S_CLUSTER_NAME
        ),
        _constants.NAMESPACE_NAME: MapConfig(
            ResourceAttributes.K8S_NAMESPACE_NAME
        ),
        _constants.POD_NAME: MapConfig(ResourceAttributes.K8S_POD_NAME),
    },
    _constants.K8S_NODE: {
        _constants.LOCATION: MapConfig(
            ResourceAttributes.CLOUD_AVAILABILITY_ZONE,
            ResourceAttributes.CLOUD_REGION,
        ),
        _constants.CLUSTER_NAME: MapConfig(
            ResourceAttributes.K8S_CLUSTER_NAME
        ),
        _constants.NODE_NAME: MapConfig(ResourceAttributes.K8S_NODE_NAME),
    },
    _constants.K8S_CLUSTER: {
        _constants.LOCATION: MapConfig(
            ResourceAttributes.CLOUD_AVAILABILITY_ZONE,
            ResourceAttributes.CLOUD_REGION,
        ),
        _constants.CLUSTER_NAME: MapConfig(
            ResourceAttributes.K8S_CLUSTER_NAME
        ),
    },
    _constants.AWS_EC2_INSTANCE: {
        _constants.INSTANCE_ID: MapConfig(ResourceAttributes.HOST_ID),
        _constants.REGION: MapConfig(
            ResourceAttributes.CLOUD_AVAILABILITY_ZONE,
            ResourceAttributes.CLOUD_REGION,
        ),
        _constants.AWS_ACCOUNT: MapConfig(ResourceAttributes.CLOUD_ACCOUNT_ID),
    },
    _constants.GENERIC_TASK: {
        _constants.LOCATION: MapConfig(
            ResourceAttributes.CLOUD_AVAILABILITY_ZONE,
            ResourceAttributes.CLOUD_REGION,
            fallback="global",
        ),
        _constants.NAMESPACE: MapConfig(ResourceAttributes.SERVICE_NAMESPACE),
        _constants.JOB: MapConfig(
            ResourceAttributes.SERVICE_NAME,
            ResourceAttributes.FAAS_NAME,
        ),
        _constants.TASK_ID: MapConfig(
            ResourceAttributes.SERVICE_INSTANCE_ID,
            ResourceAttributes.FAAS_INSTANCE,
        ),
    },
    _constants.GENERIC_NODE: {
        _constants.LOCATION: MapConfig(
            ResourceAttributes.CLOUD_AVAILABILITY_ZONE,
            ResourceAttributes.CLOUD_REGION,
            fallback="global",
        ),
        _constants.NAMESPACE: MapConfig(ResourceAttributes.SERVICE_NAMESPACE),
        _constants.NODE_ID: MapConfig(
            ResourceAttributes.HOST_ID, ResourceAttributes.HOST_NAME
        ),
    },
}


@dataclass
class MonitoredResourceData:
    """Dataclass representing a protobuf monitored resource. Make sure to convert to a protobuf
    if needed."""

    type: str
    labels: Mapping[str, str]


def get_monitored_resource(
    resource: Resource,
) -> Optional[MonitoredResourceData]:
    """Add Google resource specific information (e.g. instance id, region).

    See
    https://cloud.google.com/monitoring/custom-metrics/creating-metrics#custom-metric-resources
    for supported types
    Args:
            resource: OTel resource
    """

    attrs = resource.attributes

    platform = attrs.get(ResourceAttributes.CLOUD_PLATFORM_KEY)
    if platform == ResourceAttributes.GCP_COMPUTE_ENGINE:
        mr = _create_monitored_resource(_constants.GCE_INSTANCE, attrs)
    elif platform == ResourceAttributes.GCP_KUBERNETES_ENGINE:
        if ResourceAttributes.K8S_CONTAINER_NAME in attrs:
            mr = _create_monitored_resource(_constants.K8S_CONTAINER, attrs)
        elif ResourceAttributes.K8S_POD_NAME in attrs:
            mr = _create_monitored_resource(_constants.K8S_POD, attrs)
        elif ResourceAttributes.K8S_NODE_NAME in attrs:
            mr = _create_monitored_resource(_constants.K8S_NODE, attrs)
        else:
            mr = _create_monitored_resource(_constants.K8S_CLUSTER, attrs)
    elif platform == ResourceAttributes.AWS_EC2:
        mr = _create_monitored_resource(_constants.AWS_EC2_INSTANCE, attrs)
    else:
        # fallback to generic_task
        if (
            ResourceAttributes.SERVICE_NAME in attrs
            or ResourceAttributes.FAAS_NAME in attrs
        ) and (
            ResourceAttributes.SERVICE_INSTANCE_ID in attrs
            or ResourceAttributes.FAAS_INSTANCE in attrs
        ):
            mr = _create_monitored_resource(_constants.GENERIC_TASK, attrs)
        else:
            mr = _create_monitored_resource(_constants.GENERIC_NODE, attrs)

    return mr


def _create_monitored_resource(
    monitored_resource_type: str, resource_attrs: Attributes
) -> MonitoredResourceData:
    mapping = MAPPINGS[monitored_resource_type]
    labels: Dict[str, str] = {}

    for mr_key, map_config in mapping.items():
        mr_value = None
        for otel_key in map_config.otel_keys:
            if otel_key in resource_attrs and not str(
                resource_attrs[otel_key]
            ).startswith(_constants.UNKNOWN_SERVICE_PREFIX):
                mr_value = resource_attrs[otel_key]
                break

        if (
            mr_value is None
            and ResourceAttributes.SERVICE_NAME in map_config.otel_keys
        ):
            # The service name started with unknown_service, and was ignored above.
            mr_value = resource_attrs.get(ResourceAttributes.SERVICE_NAME)

        if mr_value is None:
            mr_value = map_config.fallback

        # OTel attribute values can be any of str, bool, int, float, or Sequence of any of
        # them. Encode any non-strings as json string
        if not isinstance(mr_value, str):
            mr_value = json.dumps(
                mr_value, sort_keys=True, indent=None, separators=(",", ":")
            )
        labels[mr_key] = mr_value

    return MonitoredResourceData(type=monitored_resource_type, labels=labels)
