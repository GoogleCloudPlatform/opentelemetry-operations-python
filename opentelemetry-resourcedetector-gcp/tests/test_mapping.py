# Copyright 2022 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import dataclasses

import pytest
from opentelemetry.resourcedetector.gcp_resource_detector._mapping import (
    get_monitored_resource,
)
from opentelemetry.sdk.resources import Attributes, LabelValue, Resource
from syrupy.assertion import SnapshotAssertion


@pytest.mark.parametrize(
    "otel_attributes",
    [
        # GCE
        pytest.param(
            {
                "cloud.platform": "gcp_compute_engine",
                "cloud.availability_zone": "foo",
                "host.id": "myhost",
            },
            id="gce instance",
        ),
        # k8s container
        pytest.param(
            {
                "cloud.platform": "gcp_kubernetes_engine",
                "cloud.availability_zone": "myavailzone",
                "k8s.cluster.name": "mycluster",
                "k8s.namespace.name": "myns",
                "k8s.pod.name": "mypod",
                "k8s.container.name": "mycontainer",
            },
            id="k8s container",
        ),
        pytest.param(
            {
                "cloud.platform": "gcp_kubernetes_engine",
                "cloud.region": "myregion",
                "k8s.cluster.name": "mycluster",
                "k8s.namespace.name": "myns",
                "k8s.pod.name": "mypod",
                "k8s.container.name": "mycontainer",
            },
            id="k8s container region fallback",
        ),
        # k8s pod
        pytest.param(
            {
                "cloud.platform": "gcp_kubernetes_engine",
                "cloud.availability_zone": "myavailzone",
                "k8s.cluster.name": "mycluster",
                "k8s.namespace.name": "myns",
                "k8s.pod.name": "mypod",
            },
            id="k8s pod",
        ),
        pytest.param(
            {
                "cloud.platform": "gcp_kubernetes_engine",
                "cloud.region": "myregion",
                "k8s.cluster.name": "mycluster",
                "k8s.namespace.name": "myns",
                "k8s.pod.name": "mypod",
            },
            id="k8s pod region fallback",
        ),
        # k8s node
        pytest.param(
            {
                "cloud.platform": "gcp_kubernetes_engine",
                "cloud.availability_zone": "myavailzone",
                "k8s.cluster.name": "mycluster",
                "k8s.namespace.name": "myns",
                "k8s.node.name": "mynode",
            },
            id="k8s node",
        ),
        pytest.param(
            {
                "cloud.platform": "gcp_kubernetes_engine",
                "cloud.region": "myregion",
                "k8s.cluster.name": "mycluster",
                "k8s.namespace.name": "myns",
                "k8s.node.name": "mynode",
            },
            id="k8s node region fallback",
        ),
        # k8s cluster
        pytest.param(
            {
                "cloud.platform": "gcp_kubernetes_engine",
                "cloud.availability_zone": "myavailzone",
                "k8s.cluster.name": "mycluster",
                "k8s.namespace.name": "myns",
            },
            id="k8s cluster",
        ),
        pytest.param(
            {
                "cloud.platform": "gcp_kubernetes_engine",
                "cloud.region": "myregion",
                "k8s.cluster.name": "mycluster",
                "k8s.namespace.name": "myns",
            },
            id="k8s cluster region fallback",
        ),
        # aws ec2
        pytest.param(
            {
                "cloud.platform": "aws_ec2",
                "cloud.availability_zone": "myavailzone",
                "host.id": "myhostid",
                "cloud.account.id": "myawsaccount",
            },
            id="aws ec2",
        ),
        pytest.param(
            {
                "cloud.platform": "aws_ec2",
                "cloud.region": "myregion",
                "host.id": "myhostid",
                "cloud.account.id": "myawsaccount",
            },
            id="aws ec2 region fallback",
        ),
        # generic task
        pytest.param(
            {
                "cloud.availability_zone": "myavailzone",
                "service.namespace": "servicens",
                "service.name": "servicename",
                "service.instance.id": "serviceinstanceid",
            },
            id="generic task",
        ),
        pytest.param(
            {
                "cloud.region": "myregion",
                "service.namespace": "servicens",
                "service.name": "servicename",
                "service.instance.id": "serviceinstanceid",
            },
            id="generic task fallback region",
        ),
        pytest.param(
            {
                "service.namespace": "servicens",
                "service.name": "servicename",
                "service.instance.id": "serviceinstanceid",
            },
            id="generic task fallback global",
        ),
        pytest.param(
            {
                "service.name": "unknown_service",
                "cloud.region": "myregion",
                "service.namespace": "servicens",
                "faas.name": "faasname",
                "faas.instance": "faasinstance",
            },
            id="generic task faas",
        ),
        pytest.param(
            {
                "service.name": "unknown_service",
                "cloud.region": "myregion",
                "service.namespace": "servicens",
                "faas.instance": "faasinstance",
            },
            id="generic task faas fallback",
        ),
        # generic node
        pytest.param(
            {
                "cloud.availability_zone": "myavailzone",
                "service.namespace": "servicens",
                "service.name": "servicename",
                "host.id": "hostid",
            },
            id="generic node",
        ),
        pytest.param(
            {
                "cloud.region": "myregion",
                "service.namespace": "servicens",
                "service.name": "servicename",
                "host.id": "hostid",
            },
            id="generic node fallback region",
        ),
        pytest.param(
            {
                "service.namespace": "servicens",
                "service.name": "servicename",
                "host.id": "hostid",
            },
            id="generic node fallback global",
        ),
        pytest.param(
            {
                "service.namespace": "servicens",
                "service.name": "servicename",
                "host.name": "hostname",
            },
            id="generic node fallback host name",
        ),
        # fallback empty
        pytest.param(
            {"foo": "bar", "no.useful": "resourceattribs"},
            id="fallback generic node",
        ),
        pytest.param(
            {},
            id="empty",
        ),
    ],
)
def test_get_monitored_resource(
    otel_attributes: Attributes, snapshot: SnapshotAssertion
) -> None:
    resource = Resource(otel_attributes)
    monitored_resource_data = get_monitored_resource(resource)
    as_dict = dataclasses.asdict(monitored_resource_data)
    assert as_dict == snapshot


@pytest.mark.parametrize(
    ("value", "expect"),
    [
        (None, ""),
        (123, "123"),
        (123.4, "123.4"),
        ([1, 2, 3, 4], "[1,2,3,4]"),
        ([1.1, 2.2, 3.3, 4.4], "[1.1,2.2,3.3,4.4]"),
        (["a", "b", "c", "d"], '["a","b","c","d"]'),
    ],
)
def test_non_string_values(value: LabelValue, expect: str):
    # host.id will end up in generic_node's node_id label
    monitored_resource_data = get_monitored_resource(
        Resource({"host.id": value})
    )
    assert monitored_resource_data is not None

    value_as_gcm_label = monitored_resource_data.labels["node_id"]
    assert value_as_gcm_label == expect
