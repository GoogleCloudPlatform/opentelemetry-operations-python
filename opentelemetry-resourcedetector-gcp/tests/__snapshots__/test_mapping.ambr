# name: test_get_monitored_resource[aws ec2 region fallback]
  dict({
    'labels': dict({
      'aws_account': 'myawsaccount',
      'instance_id': 'myhostid',
      'region': 'myregion',
    }),
    'type': 'aws_ec2_instance',
  })
# ---
# name: test_get_monitored_resource[aws ec2]
  dict({
    'labels': dict({
      'aws_account': 'myawsaccount',
      'instance_id': 'myhostid',
      'region': 'myavailzone',
    }),
    'type': 'aws_ec2_instance',
  })
# ---
# name: test_get_monitored_resource[empty]
  dict({
    'labels': dict({
      'location': 'global',
      'namespace': '',
      'node_id': '',
    }),
    'type': 'generic_node',
  })
# ---
# name: test_get_monitored_resource[fallback generic node]
  dict({
    'labels': dict({
      'location': 'global',
      'namespace': '',
      'node_id': '',
    }),
    'type': 'generic_node',
  })
# ---
# name: test_get_monitored_resource[gce instance]
  dict({
    'labels': dict({
      'instance_id': 'myhost',
      'zone': 'foo',
    }),
    'type': 'gce_instance',
  })
# ---
# name: test_get_monitored_resource[generic node fallback global]
  dict({
    'labels': dict({
      'location': 'global',
      'namespace': 'servicens',
      'node_id': 'hostid',
    }),
    'type': 'generic_node',
  })
# ---
# name: test_get_monitored_resource[generic node fallback host name]
  dict({
    'labels': dict({
      'location': 'global',
      'namespace': 'servicens',
      'node_id': 'hostname',
    }),
    'type': 'generic_node',
  })
# ---
# name: test_get_monitored_resource[generic node fallback region]
  dict({
    'labels': dict({
      'location': 'myregion',
      'namespace': 'servicens',
      'node_id': 'hostid',
    }),
    'type': 'generic_node',
  })
# ---
# name: test_get_monitored_resource[generic node]
  dict({
    'labels': dict({
      'location': 'myavailzone',
      'namespace': 'servicens',
      'node_id': 'hostid',
    }),
    'type': 'generic_node',
  })
# ---
# name: test_get_monitored_resource[generic task fallback global]
  dict({
    'labels': dict({
      'job': 'servicename',
      'location': 'global',
      'namespace': 'servicens',
      'task_id': 'serviceinstanceid',
    }),
    'type': 'generic_task',
  })
# ---
# name: test_get_monitored_resource[generic task faas]
  dict({
    'labels': dict({
      'job': 'faasname',
      'location': 'myregion',
      'namespace': 'servicens',
      'task_id': 'faasinstance',
    }),
    'type': 'generic_task',
  })
# ---
# name: test_get_monitored_resource[generic task faas fallback]
  dict({
    'labels': dict({
      'job': 'unknown_service',
      'location': 'myregion',
      'namespace': 'servicens',
      'task_id': 'faasinstance',
    }),
    'type': 'generic_task',
  })
# ---
# name: test_get_monitored_resource[generic task fallback region]
  dict({
    'labels': dict({
      'job': 'servicename',
      'location': 'myregion',
      'namespace': 'servicens',
      'task_id': 'serviceinstanceid',
    }),
    'type': 'generic_task',
  })
# ---
# name: test_get_monitored_resource[generic task]
  dict({
    'labels': dict({
      'job': 'servicename',
      'location': 'myavailzone',
      'namespace': 'servicens',
      'task_id': 'serviceinstanceid',
    }),
    'type': 'generic_task',
  })
# ---
# name: test_get_monitored_resource[k8s cluster region fallback]
  dict({
    'labels': dict({
      'cluster_name': 'mycluster',
      'location': 'myregion',
    }),
    'type': 'k8s_cluster',
  })
# ---
# name: test_get_monitored_resource[k8s cluster]
  dict({
    'labels': dict({
      'cluster_name': 'mycluster',
      'location': 'myavailzone',
    }),
    'type': 'k8s_cluster',
  })
# ---
# name: test_get_monitored_resource[k8s container region fallback]
  dict({
    'labels': dict({
      'cluster_name': 'mycluster',
      'container_name': 'mycontainer',
      'location': 'myregion',
      'namespace_name': 'myns',
      'pod_name': 'mypod',
    }),
    'type': 'k8s_container',
  })
# ---
# name: test_get_monitored_resource[k8s container]
  dict({
    'labels': dict({
      'cluster_name': 'mycluster',
      'container_name': 'mycontainer',
      'location': 'myavailzone',
      'namespace_name': 'myns',
      'pod_name': 'mypod',
    }),
    'type': 'k8s_container',
  })
# ---
# name: test_get_monitored_resource[k8s node region fallback]
  dict({
    'labels': dict({
      'cluster_name': 'mycluster',
      'location': 'myregion',
      'node_name': 'mynode',
    }),
    'type': 'k8s_node',
  })
# ---
# name: test_get_monitored_resource[k8s node]
  dict({
    'labels': dict({
      'cluster_name': 'mycluster',
      'location': 'myavailzone',
      'node_name': 'mynode',
    }),
    'type': 'k8s_node',
  })
# ---
# name: test_get_monitored_resource[k8s pod region fallback]
  dict({
    'labels': dict({
      'cluster_name': 'mycluster',
      'location': 'myregion',
      'namespace_name': 'myns',
      'pod_name': 'mypod',
    }),
    'type': 'k8s_pod',
  })
# ---
# name: test_get_monitored_resource[k8s pod]
  dict({
    'labels': dict({
      'cluster_name': 'mycluster',
      'location': 'myavailzone',
      'namespace_name': 'myns',
      'pod_name': 'mypod',
    }),
    'type': 'k8s_pod',
  })
# ---
